### PR TITLE
Addl metric indicating start and end times of latest run for each healthcheck

### PR DESCRIPTION
**Pull request by RaviHari**
Fixes issue #?? https://github.com/keikoproj/active-monitor/issues/32

## Proposed Changes
  - Add StartedAt and FinishedAt metrics 
  - Add StartedAt time to status object
  - Duration of the workflow is already available with current metrics: healthcheck_runtime_seconds
  
## Testing Done
  - Tested with an existing example here is the output for metrics:
image

  - Here is the status object response for kubectl describe hc/dns-healthcheck-vbmtn -n health: 
image

### CODE REVIEW

File: controllers/healthcheck_controller.go [hunk 1]
```diff
@@ -90,23 +90,50 @@ func (r *HealthCheckReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error)
 	ctx := context.Background()
 	log := r.Log.WithValues(hcKind, req.NamespacedName)
 
+	log.Info("Starting HealthCheck reconcile ...", "for this", "hc")
```
Comment (davemasselink) [inline controllers/healthcheck_controller.go]: @RaviHari, can you help me understand the thinking behind using variable number of string params passed to the `Info()` method?

I see you're doing it here, as well as at least lines 104 and 107.

File: controllers/healthcheck_controller.go [hunk 2]
```diff
@@ -90,23 +90,50 @@ func (r *HealthCheckReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error)
...
+			log.Info("Cancelling rescheduled workflow for this healthcheck due to deletion", "deleted", "deleted")
 			r.RepeatTimersByName[req.NamespacedName.Name].Stop()
 		}
 		return ctrl.Result{}, ignoreNotFound(err)
 	}
+
+	return r.execHealthCheck(ctx, req, log, healthCheck)
+}
+
+func (r *HealthCheckReconciler) execHealthCheck(ctx context.Context, req ctrl.Request, log logr.Logger, healthCheck *activemonitorv1alpha1.HealthCheck) (ctrl.Result, error) {
```
Comment (davemasselink) [inline controllers/healthcheck_controller.go]: looks like this branch may yet first need to pull in changes from other PR (now merged to `master`)

File: metrics/collector.go [hunk 3]
```diff
@@ -32,6 +32,18 @@ var (
 	},
 		[]string{hcName},
 	)
+	MonitorStartedtime = prometheus.NewGaugeVec(prometheus.GaugeOpts{
```
Comment (davemasselink) [inline metrics/collector.go]:
```suggestion
	MonitorStartedTime = prometheus.NewGaugeVec(prometheus.GaugeOpts{
```

File: metrics/collector.go [hunk 4]
```diff
@@ -32,6 +32,18 @@ var (
...
+		Name: "healthcheck_starttime",
+		Help: "Time taken for the workflow to complete.",
+	},
+		[]string{hcName},
+	)
+	MonitorFinishedtime = prometheus.NewGaugeVec(prometheus.GaugeOpts{
```
Comment (davemasselink) [inline metrics/collector.go]:
```suggestion
	MonitorFinishedTime = prometheus.NewGaugeVec(prometheus.GaugeOpts{
```

File: metrics/collector.go [hunk 5]
```diff
@@ -52,6 +64,8 @@ func NewRegistry() *prometheus.Registry {
 	promRegistry.MustRegister(MonitorSuccess)
 	promRegistry.MustRegister(MonitorError)
 	promRegistry.MustRegister(MonitorRuntime)
+	promRegistry.MustRegister(MonitorStartedtime)
```
Comment (davemasselink) [inline metrics/collector.go]:
```suggestion
	promRegistry.MustRegister(MonitorStartedTime)
```

File: metrics/collector.go [hunk 6]
```diff
@@ -52,6 +64,8 @@ func NewRegistry() *prometheus.Registry {
...
+	promRegistry.MustRegister(MonitorFinishedtime)
```
Comment (davemasselink) [inline metrics/collector.go]:
```suggestion
	promRegistry.MustRegister(MonitorFinishedTime)
```

File: controllers/healthcheck_controller.go (hunk 1 above)
Comment (RaviHari) [reply controllers/healthcheck_controller.go]: It is from an old branch fixed this..

File: controllers/healthcheck_controller.go (hunk 2 above)
Comment (RaviHari) [reply controllers/healthcheck_controller.go]: yes.. Merged it and resolved conflicts.. should be good now

**Comment by davemasselink** [comment]
looks like the CI build is failing in the `go vet ...` step after these changes, specifically seeing:

```
go vet ./...
go: downloading github.com/onsi/gomega v1.5.0
go: downloading github.com/onsi/ginkgo v1.8.0
go: downloading sigs.k8s.io/testing_frameworks v0.1.1
go: extracting github.com/onsi/gomega v1.5.0
go: extracting github.com/onsi/ginkgo v1.8.0
go: downloading github.com/hpcloud/tail v1.0.0
go: extracting github.com/hpcloud/tail v1.0.0
go: downloading gopkg.in/tomb.v1 v1.0.0-20141024135613-dd632973f1e7
go: extracting gopkg.in/tomb.v1 v1.0.0-20141024135613-dd632973f1e7
go: extracting sigs.k8s.io/testing_frameworks v0.1.1
# github.com/keikoproj/active-monitor/controllers
controllers/healthcheck_controller.go:298:5: undefined: metrics.MonitorStartedtime
controllers/healthcheck_controller.go:299:5: undefined: metrics.MonitorFinishedtime
controllers/healthcheck_controller.go:310:5: undefined: metrics.MonitorStartedtime
controllers/healthcheck_controller.go:311:5: undefined: metrics.MonitorFinishedtime
# github.com/keikoproj/active-monitor/controllers [github.com/keikoproj/active-monitor/controllers.test]
controllers/healthcheck_controller.go:298:5: undefined: metrics.MonitorStartedtime
controllers/healthcheck_controller.go:299:5: undefined: metrics.MonitorFinishedtime
controllers/healthcheck_controller.go:310:5: undefined: metrics.MonitorStartedtime
controllers/healthcheck_controller.go:311:5: undefined: metrics.MonitorFinishedtime
Makefile:42: recipe for target 'vet' failed
make: *** [vet] Error 2
The command "make test" exited with 2.
```

I'll try pulling and building myself to see if I'm seeing the same locally sometime in next day or so

**Comment by RaviHari** [comment]
> looks like the CI build is failing in the `go vet ...` step after these changes, specifically seeing:
> 
> ```
> go vet ./...
> go: downloading github.com/onsi/gomega v1.5.0
> go: downloading github.com/onsi/ginkgo v1.8.0
> go: downloading sigs.k8s.io/testing_frameworks v0.1.1
> go: extracting github.com/onsi/gomega v1.5.0
> go: extracting github.com/onsi/ginkgo v1.8.0
> go: downloading github.com/hpcloud/tail v1.0.0
> go: extracting github.com/hpcloud/tail v1.0.0
> go: downloading gopkg.in/tomb.v1 v1.0.0-20141024135613-dd632973f1e7
> go: extracting gopkg.in/tomb.v1 v1.0.0-20141024135613-dd632973f1e7
> go: extracting sigs.k8s.io/testing_frameworks v0.1.1
> # github.com/keikoproj/active-monitor/controllers
> controllers/healthcheck_controller.go:298:5: undefined: metrics.MonitorStartedtime
> controllers/healthcheck_controller.go:299:5: undefined: metrics.MonitorFinishedtime
> controllers/healthcheck_controller.go:310:5: undefined: metrics.MonitorStartedtime
> controllers/healthcheck_controller.go:311:5: undefined: metrics.MonitorFinishedtime
> # github.com/keikoproj/active-monitor/controllers [github.com/keikoproj/active-monitor/controllers.test]
> controllers/healthcheck_controller.go:298:5: undefined: metrics.MonitorStartedtime
> controllers/healthcheck_controller.go:299:5: undefined: metrics.MonitorFinishedtime
> controllers/healthcheck_controller.go:310:5: undefined: metrics.MonitorStartedtime
> controllers/healthcheck_controller.go:311:5: undefined: metrics.MonitorFinishedtime
> Makefile:42: recipe for target 'vet' failed
> make: *** [vet] Error 2
> The command "make test" exited with 2.
> ```
> 
> I'll try pulling and building myself to see if I'm seeing the same locally sometime in next day or so

I found the issue the metric names were not updated in updated controller.. checking it in..

**Comment by davemasselink** [comment]
@RaviHari, fantastic... thanks for the update. I was away most of last week and just getting caught back up on things now. This work is ready to be merged :)

**Pull request merged**